### ISSUE TITLE
UnusedLocalVariable should not add issue when used by routine inside another routine

### ISSUE THREAD

**Issue by pkantor** [integrated-application-development/sonar-delphi#263]
### Prerequisites

- [X] This bug is in SonarDelphi, not SonarQube or my Delphi code.
- [X] This bug has not already been reported.

### SonarDelphi version

1.6.0

### SonarQube version

Community Edition 9.9.5 (build 90363)

### Issue description

During destroying thread I usually wait for it to finish working and then I free it. To know if I am "waiting" during destroy, I log something (for example every second). It looks like this, I initialize a variable with the current time so I do not get the log right away, then I check if a second has passed, and if so then I log and initialize the same variable again - This is reported as code smell (UnusedLocalVariable).   

### Steps to reproduce

I've added a test case in Java **testUsedVariableInRoutineAsParameterShouldNotAddIssue** which more or less shows my problem.

### Minimal Delphi code exhibiting the issue

[UnusedLocalVariableCheckTest.zip](https://github.com/user-attachments/files/15951981/UnusedLocalVariableCheckTest.zip)

**Comment by zaneduffield** [comment]
To me this looks like a name resolution issue on `SecondsBetween`.
In your test case the call to `SecondsBetween` isn't recognised as a use of the local variable because that function is not defined (or imported).

If you define the functions then the test passes (i.e. no issues)
```java
  @Test
  void testUsedVariableInRoutineAsParameterShouldNotAddIssue() {
    CheckVerifier.newVerifier()
            .withCheck(new UnusedLocalVariableCheck())
            .onFile(
                    new DelphiTestUnitBuilder()
                            .appendImpl("function Now: TDateTime; begin end;")
                            .appendImpl("function SecondsBetween(A,B: TDateTime): Integer; begin end;")
                            .appendImpl("")
                            .appendImpl("procedure Test;")
                            .appendImpl("begin")
                            .appendImpl("  var LastLogStamp := Now;")
                            .appendImpl("  while (WaitingForThreadToFinishWork) do //Variable changed by other thread")
                            .appendImpl("    if SecondsBetween(Now, LastLogStamp) >= 1 then")
                            .appendImpl("    begin")
                            .appendImpl("      LastLogStamp := Now;")
                            .appendImpl("      WriteLn('Still waiting');")
                            .appendImpl("    end;")
                            .appendImpl("end;"))
            .verifyNoIssues();
  }
```

Do you have a more complete reproduction where you observed the issue the first time? If you have imported `System.DateUtils` and still have the issue then there's some other problem (like a name conflict on that function).

**Comment by pkantor** [comment]
I've tried to reproduce it in a new Delphi project but it ended up without this code smell. 

@zaneduffield 

> Do you have a more complete reproduction where you observed the issue the first time?

I will try to reproduce it on a larger project. It might take a day or two.

> If you have imported System.DateUtils and System.SysUtils and still have the issue then there's some other problem (like a name conflict on that function).

I will also look into that, maybe that's the case.

**Comment by pkantor** [comment]
I managed to reproduce this issue and it seems it was my invalid configuration of SonarQube...

I have some functions in a unit that is not added to the project just to the search path. While I was configuring a dummy project in SonarQube I added **Source File Inclusions** to be 
- **.dpr
- **.dpk
- **.pas

Then when I run `sonar-scanner.bat -X` I saw this message 
![image](https://github.com/integrated-application-development/sonar-delphi/assets/153601664/e30b6181-212a-4ff8-b1c0-97e4b3695841)
Unit uUtils is not linked to the project, just added to the search path.
Because this unit was not processed by SonarQube, using any function from that unit was not recognized. So if it was the only thing in a routine using local variable it ended as the UnusedLocalVariable code smell. Here is an example:

```
procedure UnusedLocalVariable;
begin
  var LocalVar := 0;
  FunctionFromUnitInSearchPath(LocalVar);
end;
```

All I had to do to fix this issue was to add `**.dproj` to **Source File Inclusions**.

**Comment by zaneduffield** [comment]
OK just a configuration issue, then.
If you haven't already found them, there are some useful documents about this and related things
1. [docs/MANUAL.md](https://github.com/integrated-application-development/sonar-delphi/blob/master/docs/MANUAL.md)
1. [docs/CONFIGURATION.md](https://github.com/integrated-application-development/sonar-delphi/blob/master/docs/CONFIGURATION.md)
